File: src/coordination/workflow.ts
/**
 * Workflow Primitive
 *
 * Workflows are multi-step processes involving multiple agents.
 * They define a sequence of steps, each handled by specific agents,
 * with automatic handoffs between steps.
 */

import { nanoid } from 'nanoid';

/**
 * Status of a workflow.
 */
export const WorkflowStatus = {
  DRAFT: 'draft',
  READY: 'ready',
  RUNNING: 'running',
  PAUSED: 'paused',
  WAITING: 'waiting',
  COMPLETED: 'completed',
  FAILED: 'failed',
  CANCELLED: 'cancelled',
} as const;

export type WorkflowStatus = (typeof WorkflowStatus)[keyof typeof WorkflowStatus];

/**
 * Status of a workflow step.
 */
export const StepStatus = {
  PENDING: 'pending',
  RUNNING: 'running',
  COMPLETED: 'completed',
  FAILED: 'failed',
  SKIPPED: 'skipped',
} as const;

export type StepStatus = (typeof StepStatus)[keyof typeof StepStatus];

/**
 * A single step in a workflow.
 */
export interface WorkflowStepData {
  id: string;
  name: string;
  description: string;
  assignedTo?: string;
  requiredCapability?: string;
  status: StepStatus;
  inputMapping: Record<string, string>;
  inputs: Record<string, unknown>;
  outputs: Record<string, unknown>;
  executedBy?: string;
  startedAt?: Date;
  completedAt?: Date;
  errorMessage: string;
  runIf?: string;
  skipOnFailure: boolean;
  maxRetries: number;
  retryCount: number;
}

export class WorkflowStep {
  readonly id: string;
  name: string;
  description: string;
  assignedTo?: string;
  requiredCapability?: string;
  status: StepStatus;
  inputMapping: Record<string, string>;
  inputs: Record<string, unknown>;
  outputs: Record<string, unknown>;
  executedBy?: string;
  startedAt?: Date;
  completedAt?: Date;
  errorMessage: string;
  runIf?: string;
  skipOnFailure: boolean;
  maxRetries: number;
  retryCount: number;

  constructor(data: Partial<WorkflowStepData> & { name: string }) {
    this.id = data.id ?? nanoid();
    this.name = data.name;
    this.description = data.description ?? '';
    this.assignedTo = data.assignedTo;
    this.requiredCapability = data.requiredCapability;
    this.status = data.status ?? StepStatus.PENDING;
    this.inputMapping = data.inputMapping ?? {};
    this.inputs = data.inputs ?? {};
    this.outputs = data.outputs ?? {};
    this.executedBy = data.executedBy;
    this.startedAt = data.startedAt;
    this.completedAt = data.completedAt;
    this.errorMessage = data.errorMessage ?? '';
    this.runIf = data.runIf;
    this.skipOnFailure = data.skipOnFailure ?? false;
    this.maxRetries = data.maxRetries ?? 0;
    this.retryCount = data.retryCount ?? 0;
  }

  get isComplete(): boolean {
    return (
      this.status === StepStatus.COMPLETED ||
      this.status === StepStatus.FAILED ||
      this.status === StepStatus.SKIPPED
    );
  }

  get duration(): number | undefined {
    if (this.startedAt && this.completedAt) {
      return this.completedAt.getTime() - this.startedAt.getTime();
    }
    return undefined;
  }

  start(agentId: string): this {
    this.status = StepStatus.RUNNING;
    this.executedBy = agentId;
    this.startedAt = new Date();
    return this;
  }

  complete(outputs?: Record<string, unknown>): this {
    this.status = StepStatus.COMPLETED;
    this.completedAt = new Date();
    if (outputs) {
      this.outputs = outputs;
    }
    return this;
  }

  fail(error: string): this {
    this.status = StepStatus.FAILED;
    this.completedAt = new Date();
    this.errorMessage = error;
    return this;
  }

  skip(): this {
    this.status = StepStatus.SKIPPED;
    this.completedAt = new Date();
    return this;
  }

  reset(): this {
    this.status = StepStatus.PENDING;
    this.startedAt = undefined;
    this.completedAt = undefined;
    this.errorMessage = '';
    this.outputs = {};
    this.retryCount += 1;
    return this;
  }

  toJSON(): WorkflowStepData {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      assignedTo: this.assignedTo,
      requiredCapability: this.requiredCapability,
      status: this.status,
      executedBy: this.executedBy,
      startedAt: this.startedAt,
      completedAt: this.completedAt,
      errorMessage: this.errorMessage,
      inputMapping: this.inputMapping,
      inputs: this.inputs,
      outputs: this.outputs,
      runIf: this.runIf,
      skipOnFailure: this.skipOnFailure,
      maxRetries: this.maxRetries,
      retryCount: this.retryCount,
    };
  }

  toString(): string {
    const statusIcons: Record<StepStatus, string> = {
      pending: '⏸️',
      running: '🔄',
      completed: '✅',
      failed: '❌',
      skipped: '⏭️',
    };
    const icon = statusIcons[this.status];
    const agent = this.executedBy ?? this.assignedTo ?? `[${this.requiredCapability}]`;
    return `${icon} ${this.name} (${agent})`;
  }
}

/**
 * A multi-step workflow executed by multiple agents.
 */
export class Workflow {
  readonly id: string;
  name: string;
  description: string;
  steps: WorkflowStep[];
  currentStepIndex: number;
  status: WorkflowStatus;
  channelId?: string;
  threadId?: string;
  triggeredBy?: string;
  inputs: Record<string, unknown>;
  outputs: Record<string, unknown>;
  readonly createdAt: Date;
  startedAt?: Date;
  completedAt?: Date;
  errorMessage: string;
  continueOnStepFailure: boolean;

  constructor(data: Partial<{
    id: string;
    name: string;
    description: string;
    steps: WorkflowStep[];
    channelId: string;
    continueOnStepFailure: boolean;
  }> & { name: string }) {
    this.id = data.id ?? nanoid();
    this.name = data.name;
    this.description = data.description ?? '';
    this.steps = data.steps ?? [];
    this.currentStepIndex = 0;
    this.status = WorkflowStatus.DRAFT;
    this.channelId = data.channelId;
    this.inputs = {};
    this.outputs = {};
    this.createdAt = new Date();
    this.errorMessage = '';
    this.continueOnStepFailure = data.continueOnStepFailure ?? false;
  }

  get isRunning(): boolean {
    return this.status === WorkflowStatus.RUNNING;
  }

  get isComplete(): boolean {
    return (
      this.status === WorkflowStatus.COMPLETED ||
      this.status === WorkflowStatus.FAILED ||
      this.status === WorkflowStatus.CANCELLED
    );
  }

  get currentStep(): WorkflowStep | undefined {
    return this.steps[this.currentStepIndex];
  }

  get progressPercent(): number {
    if (this.steps.length === 0) return 0;
    const completed = this.steps.filter((s) => s.isComplete).length;
    return Math.round((completed / this.steps.length) * 100);
  }

  get duration(): number | undefined {
    if (this.startedAt && this.completedAt) {
      return this.completedAt.getTime() - this.startedAt.getTime();
    }
    return undefined;
  }

  addStep(data: {
    name: string;
    description?: string;
    assignedTo?: string;
    requiredCapability?: string;
    inputMapping?: Record<string, string>;
    skipOnFailure?: boolean;
    maxRetries?: number;
  }): WorkflowStep {
    const step = new WorkflowStep(data);
    this.steps.push(step);
    return step;
  }

  start(triggeredBy: string, inputs?: Record<string, unknown>): this {
    this.status = WorkflowStatus.RUNNING;
    this.triggeredBy = triggeredBy;
    this.startedAt = new Date();
    this.currentStepIndex = 0;

    if (inputs) {
      this.inputs = inputs;
    }

    return this;
  }

  advance(): WorkflowStep | undefined {
    this.currentStepIndex += 1;

    if (this.currentStepIndex >= this.steps.length) {
      this.complete();
      return undefined;
    }

    return this.currentStep;
  }

  complete(outputs?: Record<string, unknown>): this {
    this.status = WorkflowStatus.COMPLETED;
    this.completedAt = new Date();

    if (outputs) {
      this.outputs = outputs;
    } else {
      // Collect outputs from all steps
      for (const step of this.steps) {
        this.outputs[step.name] = step.outputs;
      }
    }

    return this;
  }

  fail(error: string): this {
    this.status = WorkflowStatus.FAILED;
    this.completedAt = new Date();
    this.errorMessage = error;
    return this;
  }

  pause(): this {
    this.status = WorkflowStatus.PAUSED;
    return this;
  }

  resume(): this {
    if (this.status === WorkflowStatus.PAUSED) {
      this.status = WorkflowStatus.RUNNING;
    }
    return this;
  }

  cancel(): this {
    this.status = WorkflowStatus.CANCELLED;
    this.completedAt = new Date();
    return this;
  }

  /**
   * Get inputs for a step based on input mapping.
   */
  getStepInputs(step: WorkflowStep): Record<string, unknown> {
    const inputs = { ...step.inputs };

    for (const [targetKey, source] of Object.entries(step.inputMapping)) {
      const parts = source.split('.');

      if (parts.length === 2) {
        const [sourceName, sourceKey] = parts;

        if (sourceName === 'workflow') {
          if (sourceKey in this.inputs) {
            inputs[targetKey] = this.inputs[sourceKey];
          }
        } else {
          // Find step by name
          const sourceStep = this.steps.find((s) => s.name === sourceName);
          if (sourceStep && sourceKey in sourceStep.outputs) {
            inputs[targetKey] = sourceStep.outputs[sourceKey];
          }
        }
      }
    }

    return inputs;
  }

  getSummary(): {
    id: string;
    name: string;
    status: WorkflowStatus;
    progressPercent: number;
    currentStep: string | undefined;
    stepsCompleted: number;
    stepsTotal: number;
    duration: number | undefined;
  } {
    return {
      id: this.id,
      name: this.name,
      status: this.status,
      progressPercent: this.progressPercent,
      currentStep: this.currentStep?.name,
      stepsCompleted: this.steps.filter((s) => s.status === StepStatus.COMPLETED).length,
      stepsTotal: this.steps.length,
      duration: this.duration,
    };
  }

  toJSON() {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      status: this.status,
      steps: this.steps.map((s) => s.toJSON()),
      currentStepIndex: this.currentStepIndex,
      progressPercent: this.progressPercent,
      channelId: this.channelId,
      threadId: this.threadId,
      triggeredBy: this.triggeredBy,
      inputs: this.inputs,
      outputs: this.outputs,
      createdAt: this.createdAt,
      startedAt: this.startedAt,
      completedAt: this.completedAt,
      errorMessage: this.errorMessage,
    };
  }

  toString(): string {
    const statusIcons: Record<WorkflowStatus, string> = {
      draft: '📝',
      ready: '🟢',
      running: '🔄',
      paused: '⏸️',
      waiting: '⏳',
      completed: '✅',
      failed: '❌',
      cancelled: '🚫',
    };
    const icon = statusIcons[this.status];
    return `${icon} ${this.name} (${this.progressPercent}% - ${this.steps.length} steps)`;
  }
}

/**
 * Builder pattern for creating workflows.
 */
export class WorkflowBuilder {
  private readonly _workflow: Workflow;

  constructor(name: string, description = '') {
    this._workflow = new Workflow({ name, description });
  }

  addStep(data: {
    name: string;
    description?: string;
    assignedTo?: string;
    requiredCapability?: string;
    inputMapping?: Record<string, string>;
    skipOnFailure?: boolean;
    maxRetries?: number;
  }): this {
    this._workflow.addStep(data);
    return this;
  }

  withChannel(channelId: string): this {
    this._workflow.channelId = channelId;
    return this;
  }

  withInputs(inputs: Record<string, unknown>): this {
    this._workflow.inputs = inputs;
    return this;
  }

  continueOnFailure(): this {
    this._workflow.continueOnStepFailure = true;
    return this;
  }

  build(): Workflow {
    this._workflow.status = WorkflowStatus.READY;
    return this._workflow;
  }
}

/**
 * Manages workflows within a workspace.
 */
export class WorkflowManager {
  private readonly _workflows: Map<string, Workflow> = new Map();
  private readonly _templates: Map<string, Workflow> = new Map();

  add(workflow: Workflow): Workflow {
    this._workflows.set(workflow.id, workflow);
    return workflow;
  }

  get(workflowId: string): Workflow | undefined {
    return this._workflows.get(workflowId);
  }

  listRunning(): Workflow[] {
    return Array.from(this._workflows.values()).filter((w) => w.isRunning);
  }

  listByChannel(channelId: string): Workflow[] {
    return Array.from(this._workflows.values()).filter((w) => w.channelId === channelId);
  }

  saveTemplate(name: string, workflow: Workflow): void {
    this._templates.set(name, workflow);
  }

  createFromTemplate(templateName: string): Workflow | undefined {
    const template = this._templates.get(templateName);
    if (!template) return undefined;

    const newWorkflow = new Workflow({
      name: template.name,
      description: template.description,
      continueOnStepFailure: template.continueOnStepFailure,
    });

    for (const step of template.steps) {
      newWorkflow.addStep({
        name: step.name,
        description: step.description,
        assignedTo: step.assignedTo,
        requiredCapability: step.requiredCapability,
        inputMapping: { ...step.inputMapping },
        skipOnFailure: step.skipOnFailure,
        maxRetries: step.maxRetries,
      });
    }

    newWorkflow.status = WorkflowStatus.READY;
    return this.add(newWorkflow);
  }

  getStats(): {
    total: number;
    running: number;
    completed: number;
    failed: number;
    templates: number;
  } {
    const allWorkflows = Array.from(this._workflows.values());

    return {
      total: allWorkflows.length,
      running: allWorkflows.filter((w) => w.isRunning).length,
      completed: allWorkflows.filter((w) => w.status === WorkflowStatus.COMPLETED).length,
      failed: allWorkflows.filter((w) => w.status === WorkflowStatus.FAILED).length,
      templates: this._templates.size,
    };
  }
}

File: agent_slack/coordination/workflow.py
"""
Workflow Primitive

Workflows are multi-step processes involving multiple agents.
They define a sequence of steps, each handled by specific agents,
with automatic handoffs between steps.
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Any, Callable
from enum import Enum
import uuid


class WorkflowStatus(Enum):
    """Status of a workflow."""
    
    DRAFT = "draft"              # Being defined
    READY = "ready"              # Ready to start
    RUNNING = "running"          # Currently executing
    PAUSED = "paused"            # Temporarily paused
    WAITING = "waiting"          # Waiting for external input
    COMPLETED = "completed"      # Successfully finished
    FAILED = "failed"            # Failed with error
    CANCELLED = "cancelled"      # Manually cancelled


class StepStatus(Enum):
    """Status of a workflow step."""
    
    PENDING = "pending"          # Not yet started
    RUNNING = "running"          # Currently executing
    COMPLETED = "completed"      # Successfully finished
    FAILED = "failed"            # Failed
    SKIPPED = "skipped"          # Skipped (conditional)


@dataclass
class WorkflowStep:
    """
    A single step in a workflow.
    
    Each step is executed by an agent and can produce outputs
    that feed into subsequent steps.
    """
    
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    
    # Identity
    name: str = ""
    description: str = ""
    
    # Execution
    assigned_to: Optional[str] = None        # Specific agent ID
    required_capability: Optional[str] = None # Or any agent with this capability
    
    # Status
    status: StepStatus = StepStatus.PENDING
    
    # Input/Output
    input_mapping: dict[str, str] = field(default_factory=dict)  # step_output -> this_input
    inputs: dict = field(default_factory=dict)
    outputs: dict = field(default_factory=dict)
    
    # Execution details
    executed_by: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: str = ""
    
    # Conditions
    run_if: Optional[str] = None             # Condition expression
    skip_on_failure: bool = False            # Continue workflow if this fails
    
    # Retry
    max_retries: int = 0
    retry_count: int = 0
    
    @property
    def is_complete(self) -> bool:
        return self.status in (StepStatus.COMPLETED, StepStatus.FAILED, StepStatus.SKIPPED)
    
    @property
    def duration(self) -> Optional[float]:
        if self.started_at and self.completed_at:
            return (self.completed_at - self.started_at).total_seconds()
        return None
    
    def start(self, agent_id: str) -> "WorkflowStep":
        """Mark step as started."""
        self.status = StepStatus.RUNNING
        self.executed_by = agent_id
        self.started_at = datetime.utcnow()
        return self
    
    def complete(self, outputs: dict = None) -> "WorkflowStep":
        """Mark step as completed."""
        self.status = StepStatus.COMPLETED
        self.completed_at = datetime.utcnow()
        if outputs:
            self.outputs = outputs
        return self
    
    def fail(self, error: str) -> "WorkflowStep":
        """Mark step as failed."""
        self.status = StepStatus.FAILED
        self.completed_at = datetime.utcnow()
        self.error_message = error
        return self
    
    def skip(self) -> "WorkflowStep":
        """Mark step as skipped."""
        self.status = StepStatus.SKIPPED
        self.completed_at = datetime.utcnow()
        return self
    
    def reset(self) -> "WorkflowStep":
        """Reset step for retry."""
        self.status = StepStatus.PENDING
        self.started_at = None
        self.completed_at = None
        self.error_message = ""
        self.outputs = {}
        self.retry_count += 1
        return self
    
    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "assigned_to": self.assigned_to,
            "required_capability": self.required_capability,
            "status": self.status.value,
            "executed_by": self.executed_by,
            "started_at": self.started_at.isoformat() if self.started_at else None,
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "duration": self.duration,
            "error_message": self.error_message,
            "inputs": self.inputs,
            "outputs": self.outputs,
        }
    
    def __str__(self) -> str:
        status_icons = {
            StepStatus.PENDING: "⏸️",
            StepStatus.RUNNING: "🔄",
            StepStatus.COMPLETED: "✅",
            StepStatus.FAILED: "❌",
            StepStatus.SKIPPED: "⏭️",
        }
        icon = status_icons.get(self.status, "❓")
        agent = self.executed_by or self.assigned_to or f"[{self.required_capability}]"
        return f"{icon} {self.name} ({agent})"


@dataclass
class Workflow:
    """
    A multi-step workflow executed by multiple agents.
    
    Key concepts:
    - Workflows define a sequence of steps
    - Each step can be assigned to an agent or capability
    - Outputs from one step can feed into the next
    - Workflows track overall progress and can be paused/resumed
    
    Design decisions:
    - Workflows are declarative (define what, not how)
    - Step execution is handled by the workflow engine
    - Supports both linear and conditional flows
    - Can be triggered manually or by events
    """
    
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    
    # Identity
    name: str = ""
    description: str = ""
    
    # Steps
    steps: list[WorkflowStep] = field(default_factory=list)
    current_step_index: int = 0
    
    # Status
    status: WorkflowStatus = WorkflowStatus.DRAFT
    
    # Context
    channel_id: Optional[str] = None
    thread_id: Optional[str] = None       # Discussion thread for workflow
    triggered_by: Optional[str] = None    # Agent or event that started it
    
    # Global inputs/outputs
    inputs: dict = field(default_factory=dict)
    outputs: dict = field(default_factory=dict)
    
    # Timing
    created_at: datetime = field(default_factory=datetime.utcnow)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    
    # Error handling
    error_message: str = ""
    continue_on_step_failure: bool = False
    
    @property
    def is_running(self) -> bool:
        return self.status == WorkflowStatus.RUNNING
    
    @property
    def is_complete(self) -> bool:
        return self.status in (
            WorkflowStatus.COMPLETED,
            WorkflowStatus.FAILED,
            WorkflowStatus.CANCELLED,
        )
    
    @property
    def current_step(self) -> Optional[WorkflowStep]:
        if 0 <= self.current_step_index < len(self.steps):
            return self.steps[self.current_step_index]
        return None
    
    @property
    def progress_percent(self) -> int:
        if not self.steps:
            return 0
        completed = sum(1 for s in self.steps if s.is_complete)
        return int((completed / len(self.steps)) * 100)
    
    @property
    def duration(self) -> Optional[float]:
        if self.started_at and self.completed_at:
            return (self.completed_at - self.started_at).total_seconds()
        return None
    
    def add_step(
        self,
        name: str,
        description: str = "",
        assigned_to: Optional[str] = None,
        required_capability: Optional[str] = None,
        **kwargs,
    ) -> WorkflowStep:
        """Add a step to the workflow."""
        step = WorkflowStep(
            name=name,
            description=description,
            assigned_to=assigned_to,
            required_capability=required_capability,
            **kwargs,
        )
        self.steps.append(step)
        return step
    
    def start(self, triggered_by: str, inputs: dict = None) -> "Workflow":
        """Start the workflow."""
        self.status = WorkflowStatus.RUNNING
        self.triggered_by = triggered_by
        self.started_at = datetime.utcnow()
        self.current_step_index = 0
        
        if inputs:
            self.inputs = inputs
        
        return self
    
    def advance(self) -> Optional[WorkflowStep]:
        """Advance to the next step."""
        self.current_step_index += 1
        
        if self.current_step_index >= len(self.steps):
            self.complete()
            return None
        
        return self.current_step
    
    def complete(self, outputs: dict = None) -> "Workflow":
        """Mark workflow as completed."""
        self.status = WorkflowStatus.COMPLETED
        self.completed_at = datetime.utcnow()
        
        if outputs:
            self.outputs = outputs
        else:
            # Collect outputs from all steps
            for step in self.steps:
                self.outputs[step.name] = step.outputs
        
        return self
    
    def fail(self, error: str) -> "Workflow":
        """Mark workflow as failed."""
        self.status = WorkflowStatus.FAILED
        self.completed_at = datetime.utcnow()
        self.error_message = error
        return self
    
    def pause(self) -> "Workflow":
        """Pause the workflow."""
        self.status = WorkflowStatus.PAUSED
        return self
    
    def resume(self) -> "Workflow":
        """Resume a paused workflow."""
        if self.status == WorkflowStatus.PAUSED:
            self.status = WorkflowStatus.RUNNING
        return self
    
    def cancel(self) -> "Workflow":
        """Cancel the workflow."""
        self.status = WorkflowStatus.CANCELLED
        self.completed_at = datetime.utcnow()
        return self
    
    def get_step_inputs(self, step: WorkflowStep) -> dict:
        """
        Get inputs for a step based on input mapping.
        
        Resolves references to outputs from previous steps.
        """
        inputs = step.inputs.copy()
        
        for target_key, source in step.input_mapping.items():
            # Source format: "step_name.output_key" or "workflow.input_key"
            parts = source.split(".", 1)
            
            if len(parts) == 2:
                source_name, source_key = parts
                
                if source_name == "workflow":
                    if source_key in self.inputs:
                        inputs[target_key] = self.inputs[source_key]
                else:
                    # Find step by name
                    for s in self.steps:
                        if s.name == source_name and source_key in s.outputs:
                            inputs[target_key] = s.outputs[source_key]
                            break
        
        return inputs
    
    def get_summary(self) -> dict:
        """Get a summary of workflow state."""
        return {
            "id": self.id,
            "name": self.name,
            "status": self.status.value,
            "progress_percent": self.progress_percent,
            "current_step": self.current_step.name if self.current_step else None,
            "steps_completed": sum(1 for s in self.steps if s.status == StepStatus.COMPLETED),
            "steps_total": len(self.steps),
            "duration": self.duration,
        }
    
    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "status": self.status.value,
            "steps": [s.to_dict() for s in self.steps],
            "current_step_index": self.current_step_index,
            "progress_percent": self.progress_percent,
            "channel_id": self.channel_id,
            "thread_id": self.thread_id,
            "triggered_by": self.triggered_by,
            "inputs": self.inputs,
            "outputs": self.outputs,
            "created_at": self.created_at.isoformat(),
            "started_at": self.started_at.isoformat() if self.started_at else None,
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "error_message": self.error_message,
        }
    
    def __str__(self) -> str:
        status_icons = {
            WorkflowStatus.DRAFT: "📝",
            WorkflowStatus.READY: "🟢",
            WorkflowStatus.RUNNING: "🔄",
            WorkflowStatus.PAUSED: "⏸️",
            WorkflowStatus.WAITING: "⏳",
            WorkflowStatus.COMPLETED: "✅",
            WorkflowStatus.FAILED: "❌",
            WorkflowStatus.CANCELLED: "🚫",
        }
        icon = status_icons.get(self.status, "❓")
        return f"{icon} {self.name} ({self.progress_percent}% - {len(self.steps)} steps)"


class WorkflowBuilder:
    """Builder pattern for creating workflows."""
    
    def __init__(self, name: str, description: str = ""):
        self._workflow = Workflow(name=name, description=description)
    
    def add_step(
        self,
        name: str,
        description: str = "",
        assigned_to: Optional[str] = None,
        required_capability: Optional[str] = None,
        **kwargs,
    ) -> "WorkflowBuilder":
        """Add a step to the workflow."""
        self._workflow.add_step(
            name=name,
            description=description,
            assigned_to=assigned_to,
            required_capability=required_capability,
            **kwargs,
        )
        return self
    
    def with_channel(self, channel_id: str) -> "WorkflowBuilder":
        """Set the channel for the workflow."""
        self._workflow.channel_id = channel_id
        return self
    
    def with_inputs(self, inputs: dict) -> "WorkflowBuilder":
        """Set initial inputs for the workflow."""
        self._workflow.inputs = inputs
        return self
    
    def continue_on_failure(self) -> "WorkflowBuilder":
        """Allow workflow to continue if a step fails."""
        self._workflow.continue_on_step_failure = True
        return self
    
    def build(self) -> Workflow:
        """Build and return the workflow."""
        self._workflow.status = WorkflowStatus.READY
        return self._workflow


class WorkflowManager:
    """Manages workflows within a workspace."""
    
    def __init__(self):
        self._workflows: dict[str, Workflow] = {}
        self._templates: dict[str, Workflow] = {}
    
    def add(self, workflow: Workflow) -> Workflow:
        """Add a workflow."""
        self._workflows[workflow.id] = workflow
        return workflow
    
    def get(self, workflow_id: str) -> Optional[Workflow]:
        """Get a workflow by ID."""
        return self._workflows.get(workflow_id)
    
    def list_running(self) -> list[Workflow]:
        """List all running workflows."""
        return [w for w in self._workflows.values() if w.is_running]
    
    def list_by_channel(self, channel_id: str) -> list[Workflow]:
        """List workflows in a channel."""
        return [w for w in self._workflows.values() if w.channel_id == channel_id]
    
    def save_template(self, name: str, workflow: Workflow) -> None:
        """Save a workflow as a reusable template."""
        self._templates[name] = workflow
    
    def create_from_template(self, template_name: str) -> Optional[Workflow]:
        """Create a new workflow from a template."""
        template = self._templates.get(template_name)
        if not template:
            return None
        
        # Deep copy the template
        new_workflow = Workflow(
            name=template.name,
            description=template.description,
            continue_on_step_failure=template.continue_on_step_failure,
        )
        
        for step in template.steps:
            new_workflow.add_step(
                name=step.name,
                description=step.description,
                assigned_to=step.assigned_to,
                required_capability=step.required_capability,
                input_mapping=step.input_mapping.copy(),
                skip_on_failure=step.skip_on_failure,
                max_retries=step.max_retries,
            )
        
        new_workflow.status = WorkflowStatus.READY
        return self.add(new_workflow)
    
    def get_stats(self) -> dict:
        """Get workflow statistics."""
        all_workflows = list(self._workflows.values())
        
        return {
            "total": len(all_workflows),
            "running": len([w for w in all_workflows if w.is_running]),
            "completed": len([w for w in all_workflows if w.status == WorkflowStatus.COMPLETED]),
            "failed": len([w for w in all_workflows if w.status == WorkflowStatus.FAILED]),
            "templates": len(self._templates),
        }
